services/src/main/java/org/keycloak/protocol/saml/SamlService.java

```diff
@@ -734,28 +741,80 @@ protected Response newBrowserAuthentication(AuthenticationSessionModel authSessi
         return handleBrowserAuthenticationRequest(authSession, samlProtocol, isPassive, redirectToAuthentication);
     }
 
+    public RedirectBindingProtocol newRedirectBindingProtocol() {
+        return new RedirectBindingProtocol();
+    }
+
+    public PostBindingProtocol newPostBindingProtocol() {
+        return new PostBindingProtocol();
+    }
+
     /**
      */
     @GET
-    public Response redirectBinding(@QueryParam(GeneralConstants.SAML_REQUEST_KEY) String samlRequest, @QueryParam(GeneralConstants.SAML_RESPONSE_KEY) String samlResponse, @QueryParam(GeneralConstants.RELAY_STATE) String relayState, @QueryParam(GeneralConstants.SAML_ARTIFACT_KEY) String artifact) {
+    public void redirectBinding(@Suspended AsyncResponse asyncResponse, @QueryParam(GeneralConstants.SAML_REQUEST_KEY) String samlRequest, @QueryParam(GeneralConstants.SAML_RESPONSE_KEY) String samlResponse, @QueryParam(GeneralConstants.RELAY_STATE) String relayState, @QueryParam(GeneralConstants.SAML_ARTIFACT_KEY) String artifact) {
         logger.debug("SAML GET");
         CacheControlUtil.noBackButtonCacheControlHeader();
-        return new RedirectBindingProtocol().execute(samlRequest, samlResponse, relayState, artifact);
+
+        new RedirectBindingProtocol().execute(asyncResponse, samlRequest, samlResponse, relayState, artifact);
+    }
+
+    private class ClientConnectionWrapper implements ClientConnection{
```

Comment:
I remember having already seen this class somewhere in KC :-) Possible to reuse it?

Reply:
Both are not exactly the same and they are inner classes using the HttpRequest of the parent so it cannot be extracted.